packages/plugin-react/src/index.ts

```diff
@@ -361,40 +361,51 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
     }
   }
 
-  // const runtimeId = 'react/jsx-runtime'
+  const reactJsxRuntime = 'react/jsx-runtime'
+  const reactJsxDevRuntime = 'react/jsx-dev-runtime'
+  const virtualPrefix = 'virtual:'
   // Adapted from https://github.com/alloc/vite-react-jsx
   const viteReactJsx: Plugin = {
     name: 'vite:react-jsx',
     enforce: 'pre',
     config() {
       return {
         optimizeDeps: {
-          include: ['react/jsx-dev-runtime']
+          include: [reactJsxRuntime, reactJsxDevRuntime]
         }
       }
-    }
-    // TODO: this optimization may not be necesary and it is breacking esbuild+rollup compat,
-    // see https://github.com/vitejs/vite/pull/7246#discussion_r861552185
-    // We could still do the same trick and resolve to the optimized dependency here
-    /*
-    resolveId(id: string) {
-      return id === runtimeId ? id : null
     },
-    load(id: string) {
-      if (id === runtimeId) {
-        const runtimePath = resolve.sync(runtimeId, {
-          basedir: projectRoot
-        })
-        const exports = ['jsx', 'jsxs', 'Fragment']
+    resolveId(id, importer) {
+      // Resolve runtime to a virtual path to be interoped.
+      // Since the interop code re-imports `id`, we need to prevent re-setting
+      // the virtual id if the importer is already the virtual id.
+      if (
+        importer &&
+        !importer.startsWith(virtualPrefix) &&
+        (id === reactJsxRuntime || id === reactJsxDevRuntime)
+      ) {
+        return virtualPrefix + id
```

Comment:
We should resolve prefixing with `\0`, we started to do so for all internal virtual modules

Reply:
Fixed in cc346c3